jupyterhub_singleuser_profiles/profiles.py

```diff
@@ -171,13 +169,49 @@ def empty_profile(self):
 
   @classmethod
   def merge_profiles(self, profile1, profile2):
-    profile1["name"] =  ", ".join(filter(None, [profile1.get("name", ""), profile2.get("name", "")]))
-    profile1["images"] = list(set(profile1.get("images", []) + profile2.get("images", [])))
-    profile1["users"] = list(set(profile1.get("users", []) + profile2.get("users", [])))
-    profile1["env"] = {**profile1.get('env', {}), **profile2.get('env', {})}
-    profile1["resources"] = {**profile1.get('resources', {}), **profile2.get('resources', {})}
-    profile1["services"] = {**profile1.get('services', {}), **profile2.get('services', {})}
-    return profile1
+    name = ", ".join(filter(None, [profile1.get("name", ""), profile2.get("name", "")]))
+    result = {}
+    for k, v in profile1.items():
+      result[k] = self._profile_data_merge(v, profile2.get(k, type(v)()))
+
+    result["name"] = name
+    return result
+
+  @classmethod
+  def _profile_data_merge(self, a, b):
+    """merges b into a and return merged result
+
+    NOTE: tuples and arbitrary objects are not handled as it is totally ambiguous what should happen"""
+    key = None
+    # ## debug output
+    # sys.stderr.write("DEBUG: %s to %s\n" %(b,a))
+    try:
+        if a is None or isinstance(a, str) or isinstance(a, int) or isinstance(a, float):
+            # border case for first run or if a is a primitive
+            a = b
+        elif isinstance(a, list):
+            # lists can be only appended
+            if isinstance(b, list):
+                # merge lists
+                a = list(set(a + b))
```

Comment:
This will prevent you from implementing list of dicts in the user profiles since default dictionaries aren't hashable